LLama/LLamaWeights.cs

```diff
@@ -86,15 +81,43 @@ private LLamaWeights(SafeLlamaModelHandle handle)
             NativeHandle.DangerousAddRef(ref success);
         }
 
-        #region Load
         /// <summary>
-        /// Create from a "shared" handle. The `SafeLlamaModelHandle` will not be disposed and the model will not be unloaded until <b>all</b> such handles have been disposed.
+        /// Create an instance of the model using the supplied handle and metadata.
+        /// Metadata will <b>not</b> be re-read from the handle.
         /// </summary>
         /// <param name="handle"></param>
+        /// <param name="metadata"></param>
+        private LLamaWeights(SafeLlamaModelHandle handle, IReadOnlyDictionary<string, string> metadata)
+        {
+            NativeHandle = handle;
+            Metadata = metadata;
+
+            // Increment the model reference count while this weight exists.
+            // DangerousAddRef throws if it fails, so there is no need to check "success"
+            var success = false;
+            NativeHandle.DangerousAddRef(ref success);
+        }
+
+        /// <inheritdoc />
+        ~LLamaWeights()
+        {
+            // Ensure the handle is released even if user's don't explicitly call Dispose
+            Dispose();
+        }
+
+        #region Load
+        /// <summary>
+        /// Create a new instance of the model using same NativeHandle as this model.
+        /// Metadata is also copied from the existing model rather than read from the hanlde directly
+        /// The `SafeLlamaModelHandle` will not be disposed and the model will not be unloaded until <b>ALL</b> such handles have been disposed.
+        /// </summary>
         /// <returns></returns>
-        public static LLamaWeights FromSafeModelHandle(SafeLlamaModelHandle handle)
+        public LLamaWeights CloneFromHandleWithMetadata()
         {
-            return new LLamaWeights(handle);
+            var metadataClone = Metadata
```

Comment:
I don't think there's any need to clone `Metadata` here - it's an `IReadOnlyDictionary<string, string>` so it's already immutable (sorry if that was unclear before).

Reply:
reference used